File: dotnet/examples/Simulation.cs
// Simulation.cs — RecSys.AI .NET SDK demo with a self-contained mock HTTP server.
//
// Run:  dotnet run --project examples/Simulation.csproj
//       (or: dotnet script examples/Simulation.cs  if dotnet-script is installed)

using System.Net;
using System.Text;
using System.Text.Json;
using RecommendAI;

const int MockPort = 17894;

// ── Scenario data ─────────────────────────────────────────────────────────────

var movies = new[]
{
    ("movie_001", "The Matrix",                  "sci-fi",   1999, 8.7),
    ("movie_002", "Inception",                   "sci-fi",   2010, 8.8),
    ("movie_003", "Interstellar",                "sci-fi",   2014, 8.6),
    ("movie_004", "The Dark Knight",             "action",   2008, 9.0),
    ("movie_005", "Avengers: Endgame",           "action",   2019, 8.4),
    ("movie_006", "John Wick",                   "action",   2014, 7.4),
    ("movie_007", "The Shawshank Redemption",    "drama",    1994, 9.3),
    ("movie_008", "Forrest Gump",                "drama",    1994, 8.8),
    ("movie_009", "Pulp Fiction",                "thriller", 1994, 8.9),
    ("movie_010", "The Silence of the Lambs",    "thriller", 1991, 8.6),
};

var users = new[]
{
    ("alice", "Alice Johnson", "sci-fi",   28),
    ("bob",   "Bob Smith",     "action",   35),
    ("carol", "Carol White",   "drama",    42),
    ("dave",  "Dave Brown",    "sci-fi",   23),
    ("eve",   "Eve Davis",     "thriller", 31),
};

var interactions = new (string userId, string itemId, string type, double? value)[]
{
    ("alice", "movie_001", "view",     null),
    ("alice", "movie_002", "like",     null),
    ("alice", "movie_003", "purchase", null),
    ("alice", "movie_001", "rating",   9.0),
    ("bob",   "movie_004", "view",     null),
    ("bob",   "movie_005", "like",     null),
    ("bob",   "movie_006", "purchase", null),
    ("bob",   "movie_004", "rating",   8.0),
    ("carol", "movie_007", "view",     null),
    ("carol", "movie_008", "like",     null),
    ("carol", "movie_007", "purchase", null),
    ("carol", "movie_008", "rating",   9.0),
    ("dave",  "movie_001", "view",     null),
    ("dave",  "movie_002", "purchase", null),
    ("dave",  "movie_003", "rating",   8.5),
    ("eve",   "movie_009", "view",     null),
    ("eve",   "movie_010", "like",     null),
    ("eve",   "movie_009", "purchase", null),
    ("eve",   "movie_010", "rating",   8.0),
    ("alice", "movie_002", "rating",   10.0),
    ("bob",   "movie_006", "rating",   7.5),
    ("carol", "movie_009", "view",     null),
    ("dave",  "movie_004", "view",     null),
    ("eve",   "movie_002", "view",     null),
    ("alice", "movie_004", "view",     null),
};

// ── Mock server state ─────────────────────────────────────────────────────────

var usersDb      = new Dictionary<string, Dictionary<string, object?>>();
var itemsDb      = new Dictionary<string, Dictionary<string, object?>>();
var interactsDb  = new List<Dictionary<string, object?>>();
var dbLock       = new ReaderWriterLockSlim();

static string SerializeEnum(InteractionType t) => t switch
{
    InteractionType.CartAdd    => "cart_add",
    InteractionType.CartRemove => "cart_remove",
    _                          => t.ToString().ToLowerInvariant(),
};

// ── Start mock server ─────────────────────────────────────────────────────────

var listener = new HttpListener();
listener.Prefixes.Add($"http://localhost:{MockPort}/");
listener.Start();
Console.WriteLine($"[mock] Server listening on port {MockPort}\n");

var cts = new CancellationTokenSource();
var serverTask = Task.Run(async () =>
{
    while (!cts.Token.IsCancellationRequested)
    {
        HttpListenerContext ctx;
        try { ctx = await listener.GetContextAsync().ConfigureAwait(false); }
        catch { break; }
        _ = Task.Run(() => HandleRequest(ctx));
    }
});

// ── Run simulation ─────────────────────────────────────────────────────────────

using var client = new RecommendAIClient("sim-api-key", $"http://localhost:{MockPort}");

await RunSimulationAsync(client);

cts.Cancel();
listener.Stop();
await serverTask.ConfigureAwait(false);

async Task RunSimulationAsync(RecommendAIClient c)
{
    Banner("RecSys.AI .NET SDK — Movie Streaming Simulation");

    // Step 1: seed catalogue
    Step(1, "Seeding Movie Catalogue");
    foreach (var (id, title, genre, year, rating) in movies)
    {
        var item = await c.Items.CreateAsync(id, new() {
            ["title"] = title, ["genre"] = genre, ["year"] = year, ["rating"] = rating,
        });
        Console.WriteLine($"  Created item: {item.ItemId} ({item.Properties?["title"]})");
    }
    Console.WriteLine($"  {movies.Length} movies added to catalogue.\n");

    // Step 2: register users
    Step(2, "Registering Users");
    foreach (var (id, name, genre, age) in users)
    {
        var user = await c.Users.CreateAsync(id, new() {
            ["name"] = name, ["age"] = age, ["preferred_genre"] = genre,
        });
        Console.WriteLine($"  Registered: {user.UserId} ({user.Properties?["name"]})");
    }
    Console.WriteLine($"  {users.Length} users registered.\n");

    // Step 3: record interactions
    Step(3, "Recording Watch History & Ratings");
    foreach (var (uid, iid, itype, val) in interactions)
    {
        var t = itype switch
        {
            "view"      => InteractionType.View,
            "like"      => InteractionType.Like,
            "purchase"  => InteractionType.Purchase,
            "rating"    => InteractionType.Rating,
            "click"     => InteractionType.Click,
            _           => InteractionType.View,
        };
        await c.Interactions.CreateAsync(uid, iid, t, val);
    }
    Console.WriteLine($"  {interactions.Length} interactions recorded.\n");

    // Step 4: personalised recommendations
    Step(4, "Getting Personalised Recommendations");
    foreach (var (uid, _, _, _) in users)
    {
        var recs = await c.Recommendations.GetAsync(uid, 5);
        Console.WriteLine($"  Recommendations for {uid}:");
        for (int i = 0; i < recs.Count; i++)
        {
            var r = recs[i];
            var meta = r.Metadata;
            var title = meta != null && meta.TryGetValue("title", out var t) ? t?.ToString() : r.ItemId;
            Console.WriteLine($"    {i+1}. {title,-36} (score: {r.Score:F4})  {r.Reason}");
        }
        Console.WriteLine();
    }

    // Step 5: update item
    Step(5, "Updating Item Metadata");
    var updated = await c.Items.UpdateAsync("movie_001", new() {
        ["title"] = "The Matrix", ["genre"] = "sci-fi", ["year"] = 1999,
        ["rating"] = 8.7, ["remastered"] = true,
    });
    Console.WriteLine($"  Updated movie_001 — remastered: {updated.Properties?["remastered"]}\n");

    // Step 6: update user
    Step(6, "Updating User Profile");
    var alice = await c.Users.UpdateAsync("alice", new() {
        ["name"] = "Alice Johnson", ["age"] = 29,
        ["preferred_genre"] = "sci-fi", ["subscription"] = "premium",
    });
    Console.WriteLine($"  alice subscription → {alice.Properties?["subscription"]}\n");

    // Step 7: retrieve item
    Step(7, "Verifying Item Retrieval");
    var retrieved = await c.Items.GetAsync("movie_004");
    Console.WriteLine($"  Retrieved: {retrieved.ItemId} — {retrieved.Properties?["title"]} ({retrieved.Properties?["genre"]})\n");

    // Step 8: parallel recommendations
    Step(8, "Parallel Recommendations (Task.WhenAll)");
    var parallelTasks = new[] { "alice", "bob", "carol" }
        .Select(uid => c.Recommendations.GetAsync(uid, 3))
        .ToArray();
    var allRecs = await Task.WhenAll(parallelTasks);
    var parallelUsers = new[] { "alice", "bob", "carol" };
    for (int i = 0; i < parallelUsers.Length; i++)
        Console.WriteLine($"  {parallelUsers[i]} got {allRecs[i].Count} recommendations");
    Console.WriteLine();

    // Step 9: error handling
    Step(9, "Error Handling Demo");
    try
    {
        await c.Users.GetAsync("ghost_999");
        Console.WriteLine("  ERROR: expected NotFoundException was not thrown!");
    }
    catch (NotFoundException ex)
    {
        Console.WriteLine($"  Caught NotFoundException: {ex.Message}\n");
    }

    // Step 10: cleanup
    Step(10, "Cleanup");
    await c.Users.DeleteAsync("dave");
    Console.WriteLine("  Deleted user 'dave'.");
    try
    {
        await c.Users.GetAsync("dave");
    }
    catch (NotFoundException)
    {
        Console.WriteLine("  Confirmed: 'dave' no longer exists.");
    }

    Console.WriteLine();
    Banner("Simulation complete — all steps passed!");
}

// ── Mock server request handler ───────────────────────────────────────────────

void HandleRequest(HttpListenerContext ctx)
{
    var req  = ctx.Request;
    var resp = ctx.Response;
    var path = req.Url?.AbsolutePath ?? "/";
    var method = req.HttpMethod;

    resp.ContentType = "application/json";

    try
    {
        if (path == "/api/users" && method == "POST")        { HandleUserCreate(req, resp); return; }
        if (path.StartsWith("/api/users/") && method == "GET")    { HandleUserGet(path, resp); return; }
        if (path.StartsWith("/api/users/") && method == "PUT")    { HandleUserUpdate(path, req, resp); return; }
        if (path.StartsWith("/api/users/") && method == "DELETE") { HandleUserDelete(path, resp); return; }
        if (path == "/api/items" && method == "POST")        { HandleItemCreate(req, resp); return; }
        if (path.StartsWith("/api/items/") && method == "GET")    { HandleItemGet(path, resp); return; }
        if (path.StartsWith("/api/items/") && method == "PUT")    { HandleItemUpdate(path, req, resp); return; }
        if (path.StartsWith("/api/items/") && method == "DELETE") { HandleItemDelete(path, resp); return; }
        if (path == "/api/interactions" && method == "POST") { HandleInteractionCreate(req, resp); return; }
        if (path == "/api/recommendations" && method == "GET") { HandleRecommendations(req, resp); return; }

        Write(resp, 404, JsonSerializer.Serialize(new { detail = "Not found" }));
    }
    catch (Exception ex)
    {
        Write(resp, 500, JsonSerializer.Serialize(new { detail = ex.Message }));
    }
}

string ReadBody(HttpListenerRequest req)
{
    using var sr = new StreamReader(req.InputStream, Encoding.UTF8);
    return sr.ReadToEnd();
}

Dictionary<string, object?> ParseBody(HttpListenerRequest req)
    => JsonSerializer.Deserialize<Dictionary<string, object?>>(ReadBody(req)) ?? [];

void Write(HttpListenerResponse resp, int status, string body)
{
    resp.StatusCode = status;
    var bytes = Encoding.UTF8.GetBytes(body);
    resp.ContentLength64 = bytes.Length;
    resp.OutputStream.Write(bytes);
    resp.OutputStream.Close();
}

void HandleUserCreate(HttpListenerRequest req, HttpListenerResponse resp)
{
    var body   = ParseBody(req);
    var userId = body["user_id"]?.ToString() ?? "";
    var record = new Dictionary<string, object?> {
        ["user_id"]    = userId,
        ["properties"] = body.GetValueOrDefault("properties"),
        ["created_at"] = DateTimeOffset.UtcNow,
        ["updated_at"] = DateTimeOffset.UtcNow,
    };
    dbLock.EnterWriteLock();
    usersDb[userId] = record;
    dbLock.ExitWriteLock();
    Write(resp, 201, JsonSerializer.Serialize(record));
}

void HandleUserGet(string path, HttpListenerResponse resp)
{
    var id = path["/api/users/".Length..];
    dbLock.EnterReadLock();
    var found = usersDb.TryGetValue(id, out var u);
    dbLock.ExitReadLock();
    if (!found) { Write(resp, 404, JsonSerializer.Serialize(new { detail = $"User not found: {id}" })); return; }
    Write(resp, 200, JsonSerializer.Serialize(u));
}

void HandleUserUpdate(string path, HttpListenerRequest req, HttpListenerResponse resp)
{
    var id   = path["/api/users/".Length..];
    var body = ParseBody(req);
    dbLock.EnterWriteLock();
    if (!usersDb.TryGetValue(id, out var existing))
    {
        dbLock.ExitWriteLock();
        Write(resp, 404, JsonSerializer.Serialize(new { detail = $"User not found: {id}" })); return;
    }
    existing["properties"] = body.GetValueOrDefault("properties");
    existing["updated_at"] = DateTimeOffset.UtcNow;
    dbLock.ExitWriteLock();
    Write(resp, 200, JsonSerializer.Serialize(existing));
}

void HandleUserDelete(string path, HttpListenerResponse resp)
{
    var id = path["/api/users/".Length..];
    dbLock.EnterWriteLock();
    var removed = usersDb.Remove(id);
    dbLock.ExitWriteLock();
    if (!removed) { Write(resp, 404, JsonSerializer.Serialize(new { detail = $"User not found: {id}" })); return; }
    resp.StatusCode = 204;
    resp.OutputStream.Close();
}

void HandleItemCreate(HttpListenerRequest req, HttpListenerResponse resp)
{
    var body   = ParseBody(req);
    var itemId = body["item_id"]?.ToString() ?? "";
    var record = new Dictionary<string, object?> {
        ["item_id"]    = itemId,
        ["properties"] = body.GetValueOrDefault("properties"),
        ["created_at"] = DateTimeOffset.UtcNow,
        ["updated_at"] = DateTimeOffset.UtcNow,
    };
    dbLock.EnterWriteLock();
    itemsDb[itemId] = record;
    dbLock.ExitWriteLock();
    Write(resp, 201, JsonSerializer.Serialize(record));
}

void HandleItemGet(string path, HttpListenerResponse resp)
{
    var id = path["/api/items/".Length..];
    dbLock.EnterReadLock();
    var found = itemsDb.TryGetValue(id, out var item);
    dbLock.ExitReadLock();
    if (!found) { Write(resp, 404, JsonSerializer.Serialize(new { detail = $"Item not found: {id}" })); return; }
    Write(resp, 200, JsonSerializer.Serialize(item));
}

void HandleItemUpdate(string path, HttpListenerRequest req, HttpListenerResponse resp)
{
    var id   = path["/api/items/".Length..];
    var body = ParseBody(req);
    dbLock.EnterWriteLock();
    if (!itemsDb.TryGetValue(id, out var existing))
    {
        dbLock.ExitWriteLock();
        Write(resp, 404, JsonSerializer.Serialize(new { detail = $"Item not found: {id}" })); return;
    }
    existing["properties"] = body.GetValueOrDefault("properties");
    existing["updated_at"] = DateTimeOffset.UtcNow;
    dbLock.ExitWriteLock();
    Write(resp, 200, JsonSerializer.Serialize(existing));
}

void HandleItemDelete(string path, HttpListenerResponse resp)
{
    var id = path["/api/items/".Length..];
    dbLock.EnterWriteLock();
    var removed = itemsDb.Remove(id);
    dbLock.ExitWriteLock();
    if (!removed) { Write(resp, 404, JsonSerializer.Serialize(new { detail = $"Item not found: {id}" })); return; }
    resp.StatusCode = 204;
    resp.OutputStream.Close();
}

void HandleInteractionCreate(HttpListenerRequest req, HttpListenerResponse resp)
{
    var body = ParseBody(req);
    body["interaction_id"] = $"ia_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}";
    body["timestamp"]      = DateTimeOffset.UtcNow;
    dbLock.EnterWriteLock();
    interactsDb.Add(body);
    dbLock.ExitWriteLock();
    Write(resp, 201, JsonSerializer.Serialize(body));
}

void HandleRecommendations(HttpListenerRequest req, HttpListenerResponse resp)
{
    var userId = req.QueryString["user_id"] ?? "";
    int.TryParse(req.QueryString["limit"] ?? "10", out int limit);
    var recs = ComputeRecommendations(userId, limit);
    Write(resp, 200, JsonSerializer.Serialize(new { user_id = userId, recommendations = recs }));
}

List<Dictionary<string, object?>> ComputeRecommendations(string userId, int limit)
{
    dbLock.EnterReadLock();
    var seen = new HashSet<string>(interactsDb
        .Where(ia => ia["user_id"]?.ToString() == userId)
        .Select(ia => ia["item_id"]?.ToString() ?? ""));

    string preferredGenre = "";
    if (usersDb.TryGetValue(userId, out var u))
    {
        var props = u["properties"] as Dictionary<string, object?> ??
                    JsonSerializer.Deserialize<Dictionary<string, object?>>(
                        JsonSerializer.Serialize(u["properties"])) ?? [];
        props.TryGetValue("preferred_genre", out var pg);
        preferredGenre = pg?.ToString() ?? "";
    }

    var candidates = new List<(double score, Dictionary<string, object?> rec)>();
    foreach (var (itemId, item) in itemsDb)
    {
        if (seen.Contains(itemId)) continue;
        var props = item["properties"] as Dictionary<string, object?> ??
                    JsonSerializer.Deserialize<Dictionary<string, object?>>(
                        JsonSerializer.Serialize(item["properties"])) ?? [];
        props.TryGetValue("rating",  out var rawRating);
        props.TryGetValue("genre",   out var rawGenre);
        props.TryGetValue("title",   out var rawTitle);
        double rating = rawRating switch
        {
            JsonElement je => je.GetDouble(),
            double d       => d,
            _              => 5.0,
        };
        var genre = rawGenre?.ToString() ?? "";
        double score = rating / 10.0;
        if (!string.IsNullOrEmpty(preferredGenre) && genre == preferredGenre) score += 0.2;
        score += Djb2(userId + itemId) % 100 / 1000.0;
        score  = Math.Min(score, 1.0);
        score  = Math.Round(score, 4);
        string reason = preferredGenre != "" && genre == preferredGenre
            ? $"Matches preferred genre: {preferredGenre}"
            : "Highly rated content";
        candidates.Add((score, new() {
            ["item_id"]  = itemId,
            ["score"]    = score,
            ["reason"]   = reason,
            ["metadata"] = new Dictionary<string, object?> { ["title"] = rawTitle?.ToString() },
        }));
    }
    dbLock.ExitReadLock();

    return candidates
        .OrderByDescending(c => c.score)
        .Take(limit)
        .Select(c => c.rec)
        .ToList();
}

static ulong Djb2(string s)
{
    ulong h = 5381;
    foreach (char c in s) h = ((h << 5) + h) + c;
    return h;
}

// ── Display helpers ───────────────────────────────────────────────────────────

static void Banner(string text)
{
    var line = new string('=', text.Length + 4);
    Console.WriteLine(line);
    Console.WriteLine($"= {text} =");
    Console.WriteLine(line);
    Console.WriteLine();
}

static void Step(int n, string title) => Console.WriteLine($"── Step {n}: {title}");

File: php/examples/simulation.php
<?php

declare(strict_types=1);
/**
 * RecSys.AI PHP SDK — Movie Streaming Simulation
 *
 * Starts a PHP built-in HTTP server running mock_server.php on port 17896,
 * then exercises the full SDK without any live service.
 *
 * Run:  php examples/simulation.php
 */

require __DIR__ . '/../vendor/autoload.php';

use RecommendAI\RecommendAIClient;
use RecommendAI\InteractionType;
use RecommendAI\NotFoundException;
use RecommendAI\RecommendAIException;

define('MOCK_PORT', 17896);
define('MOCK_HOST', '127.0.0.1');
define('STATE_FILE', sys_get_temp_dir() . '/recommendai_php_sim_state.json');

// ── Wipe any leftover state ────────────────────────────────────────────────────
if (file_exists(STATE_FILE)) {
    unlink(STATE_FILE);
}

// ── Scenario data ──────────────────────────────────────────────────────────────

$movies = [
    ['id' => 'movie_001', 'title' => 'The Matrix',                 'genre' => 'sci-fi',   'year' => 1999, 'rating' => 8.7],
    ['id' => 'movie_002', 'title' => 'Inception',                  'genre' => 'sci-fi',   'year' => 2010, 'rating' => 8.8],
    ['id' => 'movie_003', 'title' => 'Interstellar',               'genre' => 'sci-fi',   'year' => 2014, 'rating' => 8.6],
    ['id' => 'movie_004', 'title' => 'The Dark Knight',            'genre' => 'action',   'year' => 2008, 'rating' => 9.0],
    ['id' => 'movie_005', 'title' => 'Avengers: Endgame',          'genre' => 'action',   'year' => 2019, 'rating' => 8.4],
    ['id' => 'movie_006', 'title' => 'John Wick',                  'genre' => 'action',   'year' => 2014, 'rating' => 7.4],
    ['id' => 'movie_007', 'title' => 'The Shawshank Redemption',   'genre' => 'drama',    'year' => 1994, 'rating' => 9.3],
    ['id' => 'movie_008', 'title' => 'Forrest Gump',               'genre' => 'drama',    'year' => 1994, 'rating' => 8.8],
    ['id' => 'movie_009', 'title' => 'Pulp Fiction',               'genre' => 'thriller', 'year' => 1994, 'rating' => 8.9],
    ['id' => 'movie_010', 'title' => 'The Silence of the Lambs',   'genre' => 'thriller', 'year' => 1991, 'rating' => 8.6],
];

$users = [
    ['id' => 'alice', 'name' => 'Alice Johnson', 'preferred_genre' => 'sci-fi',   'age' => 28],
    ['id' => 'bob',   'name' => 'Bob Smith',     'preferred_genre' => 'action',   'age' => 35],
    ['id' => 'carol', 'name' => 'Carol White',   'preferred_genre' => 'drama',    'age' => 42],
    ['id' => 'dave',  'name' => 'Dave Brown',    'preferred_genre' => 'sci-fi',   'age' => 23],
    ['id' => 'eve',   'name' => 'Eve Davis',     'preferred_genre' => 'thriller', 'age' => 31],
];

$interactions = [
    ['user' => 'alice', 'item' => 'movie_001', 'type' => InteractionType::VIEW,     'value' => null ],
    ['user' => 'alice', 'item' => 'movie_002', 'type' => InteractionType::LIKE,     'value' => null ],
    ['user' => 'alice', 'item' => 'movie_003', 'type' => InteractionType::PURCHASE, 'value' => null ],
    ['user' => 'alice', 'item' => 'movie_001', 'type' => InteractionType::RATING,   'value' => 9.0  ],
    ['user' => 'bob',   'item' => 'movie_004', 'type' => InteractionType::VIEW,     'value' => null ],
    ['user' => 'bob',   'item' => 'movie_005', 'type' => InteractionType::LIKE,     'value' => null ],
    ['user' => 'bob',   'item' => 'movie_006', 'type' => InteractionType::PURCHASE, 'value' => null ],
    ['user' => 'bob',   'item' => 'movie_004', 'type' => InteractionType::RATING,   'value' => 8.0  ],
    ['user' => 'carol', 'item' => 'movie_007', 'type' => InteractionType::VIEW,     'value' => null ],
    ['user' => 'carol', 'item' => 'movie_008', 'type' => InteractionType::LIKE,     'value' => null ],
    ['user' => 'carol', 'item' => 'movie_007', 'type' => InteractionType::PURCHASE, 'value' => null ],
    ['user' => 'carol', 'item' => 'movie_008', 'type' => InteractionType::RATING,   'value' => 9.0  ],
    ['user' => 'dave',  'item' => 'movie_001', 'type' => InteractionType::VIEW,     'value' => null ],
    ['user' => 'dave',  'item' => 'movie_002', 'type' => InteractionType::PURCHASE, 'value' => null ],
    ['user' => 'dave',  'item' => 'movie_003', 'type' => InteractionType::RATING,   'value' => 8.5  ],
    ['user' => 'eve',   'item' => 'movie_009', 'type' => InteractionType::VIEW,     'value' => null ],
    ['user' => 'eve',   'item' => 'movie_010', 'type' => InteractionType::LIKE,     'value' => null ],
    ['user' => 'eve',   'item' => 'movie_009', 'type' => InteractionType::PURCHASE, 'value' => null ],
    ['user' => 'eve',   'item' => 'movie_010', 'type' => InteractionType::RATING,   'value' => 8.0  ],
    ['user' => 'alice', 'item' => 'movie_002', 'type' => InteractionType::RATING,   'value' => 10.0 ],
    ['user' => 'bob',   'item' => 'movie_006', 'type' => InteractionType::RATING,   'value' => 7.5  ],
    ['user' => 'carol', 'item' => 'movie_009', 'type' => InteractionType::VIEW,     'value' => null ],
    ['user' => 'dave',  'item' => 'movie_004', 'type' => InteractionType::VIEW,     'value' => null ],
    ['user' => 'eve',   'item' => 'movie_002', 'type' => InteractionType::VIEW,     'value' => null ],
    ['user' => 'alice', 'item' => 'movie_004', 'type' => InteractionType::VIEW,     'value' => null ],
];

// ── Launch PHP built-in server ─────────────────────────────────────────────────

$docroot    = __DIR__;
$serverCmd  = sprintf(
    'php -S %s:%d %s',
    MOCK_HOST,
    MOCK_PORT,
    escapeshellarg($docroot . '/mock_server.php')
);

$descriptors = [
    0 => ['pipe', 'r'],
    1 => ['pipe', 'w'],
    2 => ['pipe', 'w'],
];
$process = proc_open($serverCmd, $descriptors, $pipes);
if (!is_resource($process)) {
    fwrite(STDERR, "Failed to start mock server.\n");
    exit(1);
}
// Wait for server to be ready
usleep(500_000);
echo "[mock] PHP built-in server started on port " . MOCK_PORT . "\n\n";

register_shutdown_function(static function () use ($process, $pipes, $state_file): void {
    foreach ($pipes as $p) { @fclose($p); }
    proc_terminate($process);
    if (file_exists(STATE_FILE)) { unlink(STATE_FILE); }
});

// ── SDK client ─────────────────────────────────────────────────────────────────

$client = new RecommendAIClient(
    apiKey:  'sim-api-key',
    baseUrl: 'http://' . MOCK_HOST . ':' . MOCK_PORT,
);

// ── Helpers ────────────────────────────────────────────────────────────────────

function banner(string $text): void
{
    $line = str_repeat('=', strlen($text) + 4);
    echo $line . "\n";
    echo "= {$text} =\n";
    echo $line . "\n\n";
}

function step(int $n, string $title): void
{
    echo "── Step {$n}: {$title}\n";
}

banner('RecSys.AI PHP SDK — Movie Streaming Simulation');

// Step 1: seed catalogue
step(1, 'Seeding Movie Catalogue');
foreach ($movies as $m) {
    $item = $client->items()->create($m['id'], [
        'title'  => $m['title'],
        'genre'  => $m['genre'],
        'year'   => $m['year'],
        'rating' => $m['rating'],
    ]);
    echo "  Created item: {$item->item_id} ({$item->properties['title']})\n";
}
echo "  " . count($movies) . " movies added to catalogue.\n\n";

// Step 2: register users
step(2, 'Registering Users');
foreach ($users as $u) {
    $user = $client->users()->create($u['id'], [
        'name'             => $u['name'],
        'age'              => $u['age'],
        'preferred_genre'  => $u['preferred_genre'],
    ]);
    echo "  Registered: {$user->user_id} ({$user->properties['name']})\n";
}
echo "  " . count($users) . " users registered.\n\n";

// Step 3: record interactions
step(3, 'Recording Watch History & Ratings');
foreach ($interactions as $ia) {
    $client->interactions()->create(
        userId:          $ia['user'],
        itemId:          $ia['item'],
        interactionType: $ia['type'],
        value:           $ia['value'],
    );
}
echo "  " . count($interactions) . " interactions recorded.\n\n";

// Step 4: personalised recommendations
step(4, 'Getting Personalised Recommendations');
foreach ($users as $u) {
    $recs = $client->recommendations()->get($u['id'], 5);
    echo "  Recommendations for {$u['id']}:\n";
    foreach ($recs as $i => $r) {
        $title = $r->metadata['title'] ?? $r->item_id;
        printf("    %d. %-36s (score: %.4f)  %s\n", $i + 1, $title, $r->score, $r->reason);
    }
    echo "\n";
}

// Step 5: update item
step(5, 'Updating Item Metadata');
$updated = $client->items()->update('movie_001', [
    'title' => 'The Matrix', 'genre' => 'sci-fi',
    'year'  => 1999, 'rating' => 8.7, 'remastered' => true,
]);
$rem = $updated->properties['remastered'] ? 'true' : 'false';
echo "  Updated movie_001 — remastered: {$rem}\n\n";

// Step 6: update user
step(6, 'Updating User Profile');
$alice = $client->users()->update('alice', [
    'name' => 'Alice Johnson', 'age' => 29,
    'preferred_genre' => 'sci-fi', 'subscription' => 'premium',
]);
echo "  alice subscription → {$alice->properties['subscription']}\n\n";

// Step 7: verify item retrieval
step(7, 'Verifying Item Retrieval');
$retrieved = $client->items()->get('movie_004');
echo "  Retrieved: {$retrieved->item_id} — {$retrieved->properties['title']} ({$retrieved->properties['genre']})\n\n";

// Step 8: error handling
step(8, 'Error Handling Demo');
try {
    $client->users()->get('ghost_999');
    echo "  ERROR: expected NotFoundException was not thrown!\n";
} catch (NotFoundException $e) {
    echo "  Caught NotFoundException: {$e->getMessage()}\n\n";
}

// Step 9: cleanup
step(9, 'Cleanup');
$client->users()->delete('dave');
echo "  Deleted user 'dave'.\n";
try {
    $client->users()->get('dave');
} catch (NotFoundException) {
    echo "  Confirmed: 'dave' no longer exists.\n";
}

echo "\n";
banner('Simulation complete — all steps passed!');
